### ISSUE TITLE
rust-format-buffer does not preserve point position

### ISSUE THREAD

**Issue by Wilfred** [rust-lang/rust-mode#162]
Suppose I have the following code, where | is the point:

``` rust
fn foo(foo_bar_baz1: i32, foo_bar_baz2: i32, foo_bar_baz3: i32, foo_bar_baz4: i32, foo_bar_baz5: i32, foo_bar_baz6: i32) {

}

fn bar() {|}
```

Calling `M-x rust-format-buffer` naively moves to the same position after formatting, putting point in the arguments to `foo`. It then calls `rust--after-revert-hook`, which moves point to the start of the buffer.

Ideally, we'd use rustfmt's `--write-mode=diff` so emacs-rustfmt could just apply the diff. Emacs' markers are smart enough to move if text is inserted before them. This is how [gofmt is integrated with Emacs](https://github.com/dominikh/go-mode.el/blob/5d53a13bd193653728e74102c81aa931b780c9a9/go-mode.el#L934).

A simple partial fix would be to move to the same line number and column number as before -- this would handle more cases correctly.

**Comment by tromey** [comment]
One possible fix might be to count non-whitespace characters before point, and then skip that many non-whitespace characters after restoring the buffer.  The idea here is that presumably the reformat operation affects whitespace but nothing else.

**Comment by tromey** [comment]
> The idea here is that presumably the reformat operation affects whitespace but nothing else.

This turns out to be incorrect; rustfmt can change comment styles.

**Comment by Dushistov** [comment]
@tromey 

I also have problem with this, I setted `rust-format-on-save` to true, and this hurted me.
Time to time my cursor  jump to begin of buffer. So I have to turn off  `rust-format-on-save` for now, and run `M-x cargo-process-fmt` time to time by hands.

May be ask `rustfmt` team for feature: give to `rustfmt` position in file, and after formating `rustfmt` return new position? `rustfmt` can works with AST tree, and calculate cursor position against this AST tree, and recalculate based on changes in this AST tree.

**Comment by rofrol** [comment]
This is my solution based on https://github.com/rust-lang/rust-mode/pull/269#issuecomment-406780226

```elisp
;; -*- lexical-binding: t; -*-

(use-package rust-mode
    :mode ("\\.lalrpop\\'" . elm-mode)
    :straight t
    :config
    (setq rust-format-on-save nil)
    (defun rust-mode-format-buffer ()
      "format rust code without jumping to beginning when doing undo"
      (interactive)
      (shell-command
       (format "rustfmt %s"
	       (shell-quote-argument (buffer-file-name))))
      (revert-buffer t t t))

    (add-hook 'rust-mode-hook
	      (lambda ()
		(add-hook 'after-save-hook #'rust-mode-format-buffer nil 'make-it-local))))
```

**Comment by brotzeit** [comment]
This will be fixed with #269
